File: backend/app/analysis/sentiment.py
"""Sentiment classification for signals."""

from __future__ import annotations

from pydantic import BaseModel, Field

from app.llm.models import LLMRequest
from app.llm.prompts import build_sentiment_prompt
from app.models.schemas import Sentiment


class SentimentResult(BaseModel):
    """Result of sentiment analysis."""

    sentiment: Sentiment = Sentiment.NEUTRAL
    confidence: float = Field(..., ge=0.0, le=1.0)
    key_phrases: list[str] = Field(default_factory=list)


async def classify_sentiment(
    text: str, provider_name: str = "openai", model: str | None = None
) -> SentimentResult:
    """Classify the sentiment of signal text."""
    from app.llm.provider import get_provider

    provider = get_provider(provider_name)
    messages = build_sentiment_prompt(text)

    request = LLMRequest(messages=messages, model=model, temperature=0.3)
    result = await provider.complete_structured(request, SentimentResult)

    return result

File: src/lib/ai/sentiment.ts
/**
 * Sentiment classification for signals.
 * Translated from backend/app/analysis/sentiment.py
 */

import { z } from "zod";
import { logger } from "@/lib/logger";
import { getProviderWithFailover } from "./provider";
import { buildSentimentPrompt } from "./prompts";
import { SentimentResultSchema, type SentimentResult, type LLMMessage } from "./types";
import type { ProviderName } from "./provider";

export async function classifySentiment(
  text: string,
  providerName: ProviderName = "openai",
  model?: string
): Promise<SentimentResult> {
  const { provider } = getProviderWithFailover(providerName);
  const messages = buildSentimentPrompt(text);

  logger.debug("analysis.sentiment.start", {
    provider: providerName,
    text_length: text.length,
  });

  const result = await provider.completeStructured(
    messages,
    SentimentResultSchema,
    { model, temperature: 0.3 }
  );

  logger.info("analysis.sentiment.complete", {
    provider: providerName,
    sentiment: result.sentiment,
    confidence: result.confidence,
  });

  return result;
}

export async function classifySentimentWithPrompt<T extends z.ZodTypeAny>(
  messages: LLMMessage[],
  schema: T,
  providerName: ProviderName = "openai",
  temperature: number = 0.3,
  model?: string
): Promise<{ data: z.infer<T>; usage: { inputTokens: number; outputTokens: number } }> {
  const { provider } = getProviderWithFailover(providerName);

  logger.debug("analysis.sentiment.custom.start", {
    provider: providerName,
  });

  const result = await provider.completeStructuredWithUsage(messages, schema, {
    model,
    temperature,
  });

  logger.info("analysis.sentiment.custom.complete", {
    provider: providerName,
  });

  return { data: result.data as z.infer<T>, usage: result.usage };
}
